torch/_dynamo/guards.py

```diff
@@ -1247,71 +1241,105 @@ def replace(expr: str) -> Tuple[List[str], str]:
         # Only ___check_tensors knows how to return a fancy fail reason;
         # for everything else we just report the code that failed
 
         if isinstance(fail_reason, bool) and not fail_reason:
             fail_reason = part
         if isinstance(fail_reason, str):
-            reason += fail_reason
-            if config.report_all_guard_failures:
-                reason += "\n"
-            else:
+            reasons.append(fail_reason)
+            if not config.report_all_guard_failures:
                 break
 
-    if first:
-        stashed_first_fail_reason = reason
+    reason_str = "\n".join(reasons)
+    guard_failures[orig_code_map[code]].append(reason_str)
 
-    if not last:
-        return
-
-    # Technically, we're failing our last guard, which is our oldest guard due to the
-    # eval_frame.c logic that moves newest frames to head, but for logging purposes
-    # it's more useful to see the 'first' failure (if we never got a hit) since it's
-    # likely not yet been logged as a failure reason in a case of repeating failures.
-    assert stashed_first_fail_reason
-    guard_failures[orig_code_map[code]].append(stashed_first_fail_reason)
-    stashed_first_fail_reason = None
-
-    # TODO should we GuardFail our stashed_first_fail_reason too?
     try:
         if guard_fn.guard_fail_fn is not None:
             guard_fn.guard_fail_fn(
-                GuardFail(reason or "unknown reason", orig_code_map[code])
+                GuardFail(reason_str or "unknown reason", orig_code_map[code])
             )
     except Exception as e:
         log.error(
             "Failure in guard_fail_fn callback - raising here will cause a NULL Error on guard eval",
             exc_info=True,
         )
 
+    return reason_str
+
+
+def get_and_maybe_log_recompilation_reason(
+    cache_entry, frame: types.FrameType
+) -> List[str]:
+    """
+    Return the list of guard failure reasons using cache_entry.
+    If `config.report_all_guard_failures` is disabled, only the last reason
+    of the last guard is returned.
+    Logs the recompilation reason if `recompiles` logging is enabled.
+    Raises a RecompileError if `config.error_on_recompile` is enabled.
+    """
+    reasons = []
+    while cache_entry is not None:
+        reason = get_guard_fail_reason(
+            cache_entry.check_fn, cache_entry.code, frame.f_locals
+        )
+        if reason:
+            reasons.append(reason)
+        cache_entry = cache_entry.next
+
+    code = frame.f_code
+
+    if is_guard_failure_reporting_enabled() and recompiles_log.isEnabledFor(
+        logging.DEBUG
+    ):
+        if config.report_all_guard_failures:
+            failures = "\n".join(reasons)
+        else:
+            failures = reasons[-1]
+        guard_failure_details = f"triggered by the following guard failure(s):\n{textwrap.indent(failures, '- ')}"
+    else:
+        guard_failure_details = "set env var TORCH_LOGS='recompiles' to debug further"
```

Comment:
This seems dumb, we *have* all the reasons, just print it out by default

Reply:
This will only show up if recompiles is already debug, OR you're raising an error (but if you're hard erroring, might as well put as much info in as possible)